runtime/src/bank/builtin_programs.rs

```diff
@@ -62,3 +62,399 @@ mod tests {
+            feature_set::FeatureSet,
+            instruction::Instruction,
+            message::Message,
+            native_loader,
+            pubkey::Pubkey,
+            signature::Signer,
+            transaction::Transaction,
+        },
+        std::sync::Arc,
+        test_case::test_case,
+    };
+
+    enum TestPrototype<'a> {
+        Builtin(&'a BuiltinPrototype),
+        Stateless(&'a StatelessBuiltinPrototype),
+    }
+    impl<'a> TestPrototype<'a> {
+        fn deconstruct(&'a self) -> (&'a Pubkey, &'a CoreBpfMigrationConfig) {
+            match self {
+                Self::Builtin(prototype) => (
+                    &prototype.program_id,
+                    prototype.core_bpf_migration_config.as_ref().unwrap(),
+                ),
+                Self::Stateless(prototype) => (
+                    &prototype.program_id,
+                    prototype.core_bpf_migration_config.as_ref().unwrap(),
+                ),
+            }
+        }
+    }
+
+    // This test can't be used to the `compute_budget` program, unless a valid
+    // `compute_budget` program is provided as the replacement (source).
+    // See program_runtime::compute_budget_processor::process_compute_budget_instructions`.`
+    // It also can't test the `bpf_loader_upgradeable` program, as it's used in
+    // a particular way on transaction instruction processing.
+    // See `solana_svm::account_loader::load_transaction_accounts`.
+    #[test_case(TestPrototype::Builtin(&BUILTINS[0]); "system")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[1]); "vote")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[2]); "stake")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[3]); "config")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[4]); "bpf_loader_deprecated")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[5]); "bpf_loader")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[8]); "address_lookup_table")]
+    #[test_case(TestPrototype::Stateless(&STATELESS_BUILTINS[0]); "feature_gate")]
+    fn test_core_bpf_migration(prototype: TestPrototype) {
+        let (genesis_config, mint_keypair) = create_genesis_config(1_000_000 * LAMPORTS_PER_SOL);
+        let mut root_bank = Bank::new_for_tests(&genesis_config);
+
+        let (builtin_id, config) = prototype.deconstruct();
+        let feature_id = &config.feature_id;
+        let source_program_id = &config.source_program_id;
+
+        // Add the feature to the bank's inactive feature set.
+        let mut feature_set = FeatureSet::all_enabled();
+        feature_set.inactive.insert(*feature_id);
+        root_bank.feature_set = Arc::new(feature_set);
+
+        // Initialize the accounts for the source program.
+        let test_context = TestContext::new(&root_bank, builtin_id, source_program_id);
+
+        let (bank, bank_forks) = root_bank.wrap_with_bank_forks_for_tests();
+
+        // Advance one slot so that the source program becomes effective in the
```

Comment:
There are no program deployments until here, right?
Thus no need to advance the slot.

Reply:
The source program is stored using `bank.store_account`, which will cause them to be inactive for one slot. That's how you're doing it with the test you linked, too:

https://github.com/anza-xyz/agave/blob/499d36e3549abe12d3a276955aac9d3d354d7d2c/runtime/src/bank/tests.rs#L11863-L11865

Reply:
Storing it in an account != deployment. The effective slot only starts after explicit deployment through a loader, so after feature activation in this case.

Reply:
Hm, that's strange then. When I store the accounts as I'm doing, the program is loaded as `LoadedProgramType::DelayVisibility`.
https://github.com/anza-xyz/agave/blob/2c0bea045c1dd06ad931e95e5d5685ca96ccc464/programs/bpf_loader/src/lib.rs#L455-L458

Reply:
After #831, this is resolved! I've rebased onto latest tip, which includes #831.